### PR TITLE
Switch from X-Auth-{Email,Key} to Authorization for Cloudflare

### PR DESCRIPTION
**Pull request by mamercad**
https://api.cloudflare.com/#getting-started-requests

> API Tokens provide a new way to authenticate with the Cloudflare API. They allow for scoped and permissioned access to resources and use the RFC compliant Authorization Bearer Token Header.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ nosetests.xml
 octodns.egg-info/
 output/
 tmp/
+/.vscode/
```
Comment (ross) [inline .gitignore]: Please remove this change.  octoDNS avoids including editor specific ignores here. I'd recommend looking at a "global" ignores file. 

```
 ross$ grep exclude ~/.gitconfig
  excludesfile = /Users/ross/.gitignores
```

```
(env)otter:datadog-dashboards ross$ cat ~/.gitignores
*.swo
*.swp
.DS_Store
.env/
tmp/
```

That gives you a way to ignore your editor's files without us having to add all of the possible options into octoDNS.

File: octodns/provider/cloudflare.py [hunk 2]
```diff
@@ -66,16 +64,16 @@ class CloudflareProvider(BaseProvider):
     MIN_TTL = 120
     TIMEOUT = 15
 
-    def __init__(self, id, email, token, cdn=False, *args, **kwargs):
+    def __init__(self, id, token, cdn=False, *args, **kwargs):
```
Comment (ross) [inline octodns/provider/cloudflare.py]: I think we'll want to go about this in a way that doesn't break existing usage of email & token. That'd likely involve making `email` an optional parameter that defaults to `None` and then if'ing some handling details down below.

Comment (mamercad) [reply octodns/provider/cloudflare.py]: Yeah, I actually had the same thought the other day... I'll fix it up to do both.

File: .gitignore (hunk 1 above)
Comment (mamercad) [reply .gitignore]: Good call, thanks.

**Comment by parkr** [comment]
This PR LGTM. I'm going to give it a whirl with a new API token on Jekyll's DNS.

File: octodns/provider/cloudflare.py [hunk 3]
```diff
@@ -66,17 +66,24 @@ class CloudflareProvider(BaseProvider):
...
+        if email and token:
+            sess.headers.update({
+                'X-Auth-Email': email,
+                'X-Auth-Key': token,
+            })
+        else:
+            # https://api.cloudflare.com/#getting-started-requests
+            # https://tools.ietf.org/html/rfc6750#section-2.1
+            sess.headers.update({
+                'Authorization': 'Bearer %s'.format(token),
```
Comment (parkr) [inline octodns/provider/cloudflare.py]: I received this error when running with `token: env/CLOUDFLARE_TOKEN` and `CLOUDFLARE_TOKEN` set to `Read Zone.Zone` and `Edit Zone.DNS`.

```text
Traceback (most recent call last):
  File "/usr/local/bin/octodns-sync", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/octodns/cmds/sync.py", line 39, in main
    dry_run=not args.doit, force=args.force)
  File "/usr/local/lib/python2.7/site-packages/octodns/manager.py", line 301, in sync
    plans = [p for f in futures for p in f.result()]
  File "/usr/local/lib/python2.7/site-packages/octodns/manager.py", line 56, in result
    return self.func(*self.args, **self.kwargs)
  File "/usr/local/lib/python2.7/site-packages/octodns/manager.py", line 237, in _populate_and_plan
    plan = target.plan(zone)
  File "/usr/local/lib/python2.7/site-packages/octodns/provider/base.py", line 49, in plan
    exists = self.populate(existing, target=True, lenient=True)
  File "/usr/local/lib/python2.7/site-packages/octodns/provider/cloudflare.py", line 261, in populate
    records = self.zone_records(zone)
  File "/usr/local/lib/python2.7/site-packages/octodns/provider/cloudflare.py", line 214, in zone_records
    zone_id = self.zones.get(zone.name, False)
  File "/usr/local/lib/python2.7/site-packages/octodns/provider/cloudflare.py", line 114, in zones
    resp = self._request('GET', '/zones', params={'page': page})
  File "/usr/local/lib/python2.7/site-packages/octodns/provider/cloudflare.py", line 101, in _request
    raise CloudflareError(resp.json())
octodns.provider.cloudflare.CloudflareError: Invalid request headers
```

**Comment by jrenggli** [comment]
Thanks for this fix. Tested it and works for me.
What has to be done to get this merged?

File: octodns/provider/cloudflare.py (hunk 3 above)
Comment (parkr) [reply octodns/provider/cloudflare.py]: @mamercad What scopes are necessary for the token?

Comment (jrenggli) [reply octodns/provider/cloudflare.py]: I had to grant zone resources too. See screenshot below
Screenshot from 2020-02-05 16-31-17

Currently it's not possible to only use a single zone granted in Cloudflare as access to `/zones` in API is restricted.

**Comment by ross** [comment]
> What has to be done to get this merged?

Looks like there's some python2/3 issues since this PR predates that work. After that if CI is passing we'd need to make a pass through the code for final review and ship if it's there.

File: octodns/provider/cloudflare.py (hunk 3 above)
Comment (parkr) [inline octodns/provider/cloudflare.py]: @mamercad This wasn't working for me! It turns out I had to do this:

```suggestion
                'Authorization': 'Bearer {}'.format(token),
```

Using Python 2.7.17.

Comment (mamercad) [reply octodns/provider/cloudflare.py]: Thanks, updated...

**Comment by parkr** [comment]
I'd really love to see this merged, but I'm afraid I'm not sure how to approach the Python 3 CI errors. @mamercad, are you able to grok [the failures](https://github.com/github/octodns/pull/402/checks?check_run_id=428278145)? They appear to come from the new test file you added.

**Comment by mamercad** [comment]
Busy with $DAYJOB lately, I’ll try and have a look this weekend.

**Comment by ross** [comment]
High level it doesn't really make sense to copy the full Cloudflare provider tests just to test the no email case since all of the auth bits are stubbed out anyway. There were a number of Python 3 fixes in that file and looks like this PR happened prior to that. I'll push up a commit shortly that should get things working.

**Comment by parkr** [comment]
Thank you so much, @ross!

**Comment by mamercad** [comment]
Thanks much @ross!

**Pull request merged**